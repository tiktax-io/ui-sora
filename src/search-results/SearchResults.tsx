import React, { ForwardedRef, forwardRef, memo, ReactNode } from 'react'
import { useStyles } from './SearchResults.styles'
import { RowDisplayOptions, SearchResultsProps } from './SearchResults.types'

/******************************************************************************
 * @param {Array<unknown>} children array containing all items to be rendered.
 * @param {RowDisplayOptions} displayAs determines if information will be
 * displayed as cards or rows.
 * @param {ReactNode} resultsMessage optional message explaining results.
 * @param {ReactNode} searchbar component to be wrapped inside searchbar on top
 * of cards.
 * @param {ReactNode} sidebar component to be wrapped inside sidebar.
 * @returns {ReactNode} Layout containing a side div and grid.
 *****************************************************************************/
const SearchResults = memo(forwardRef(({
  children,
  resultsMessage,
  searchbar,
  sidebar
}: SearchResultsProps, ref: ForwardedRef<HTMLDivElement> | undefined) => {
  const classes = useStyles()
  return (
    <div ref={ref} className={classes.container}>
      <div className={classes.sidebar}>
        {sidebar}
      </div>
      <div className={classes.mainContent}>
        <div className={classes.searchBar}>
          {searchbar}
        </div>
        <div className={classes.resultsMessage}>
          {resultsMessage}
        </div>
        <div className={classes.grid}>
          {children.map((element, i) => (
            <div className={classes.cardContainer} key={i}>{element}</div>
          ))}
        </div>
      </div>
    </div>
  )
}))

export default SearchResults
