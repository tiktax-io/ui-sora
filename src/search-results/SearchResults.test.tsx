import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import SearchResults from './SearchResults'
import renderer from 'react-test-renderer'
import { act } from 'react-dom/test-utils'
import { SearchResultsProps } from './SearchResults.types'

describe('SearchResults layout', () => {

  /****************************************************************************
   * Reset DOM after each test
   ***************************************************************************/
  const container = document.createElement('div')
  beforeEach(() => {
    document.body.appendChild(container)
  })  
  afterEach(() => {
    document.body.removeChild(container)
  })

  test('Renders children elements', () => {
    const results = [
      'Hola',
      'Adios'
    ]
    render(<SearchResults>{results}</SearchResults>)
    const demoElement = screen.getByText(/Hola/i)
    expect(demoElement).toBeInTheDocument()
  })

  test('Results message is rendered when provided', () => {
    render(
      <SearchResults resultsMessage='I am rendered'>
        {['one','two']}
      </SearchResults>
    )
    const resultsMessage = screen.getByText(/I am rendered/i)
    expect(resultsMessage).toBeInTheDocument()
  })

  test('Searchbar component is rendered when provided', () => {
    render(
      <SearchResults searchbar={<span>searchresults searchbar here</span>}>
        {['one','two']}
      </SearchResults>
    )
    const searchBar = screen.getByText(/searchresults searchbar here/i)
    expect(searchBar).toBeInTheDocument()
  })

  test('Sidebar is rendered when provided', () => {
    render(
      <SearchResults searchbar={<span>sidebar here</span>}>
        {['one','two']}
      </SearchResults>
    )
    const sidebar = screen.getByText(/sidebar here/i)
    expect(sidebar).toBeInTheDocument()
  })

  /****************************************************************************
   * Testing responsiveness below.
   ***************************************************************************/
  const props: SearchResultsProps = {
    children: ['one', 'two'],
    displayAs: 'cards',
    resultsMessage: <span>Result message here</span>,
    searchbar: <span>searchbar here</span>,
    sidebar: <span>sidebar here</span>
  }

  test('Snapshot on desktop screens are matched', () => {
    /**************************************************************************
     * Resizing to desktop screens.
     *************************************************************************/
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1500
      })
      fireEvent.resize
    })

    const tree = renderer
      .create(
        <SearchResults {...props}/>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

  test('Snapshot on tablet screens are matched', () => {
    /**************************************************************************
     * Resizing to tablet screens.
     *************************************************************************/
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1100
      })
      fireEvent.resize
    })

    const tree = renderer
      .create(
        <SearchResults {...props}/>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

  test('Snapshot on smartphone screens are matched', () => {
    /**************************************************************************
     * Resizing to smartphone screens.
     *************************************************************************/
    const defaultWidth = window.innerWidth
    act(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 400
      })
      fireEvent.resize
    })

    const tree = renderer
      .create(
        <SearchResults {...props}/>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: defaultWidth
    })
    fireEvent.resize
  })

})
