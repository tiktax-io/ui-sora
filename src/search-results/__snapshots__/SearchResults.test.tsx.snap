// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults layout Snapshot on desktop screens are matched 1`] = `
<div
  className="container-0-2-2"
>
  <div
    className="sidebar-0-2-7"
  >
    <span>
      sidebar here
    </span>
  </div>
  <div
    className="mainContent-0-2-4"
  >
    <div
      className="searchBar-0-2-6"
    >
      <span>
        searchbar here
      </span>
    </div>
    <div
      className="resultsMessage-0-2-5"
    >
      <span>
        Result message here
      </span>
    </div>
    <div
      className="grid-0-2-3"
    >
      <div
        className="cardContainer-0-2-1"
      >
        one
      </div>
      <div
        className="cardContainer-0-2-1"
      >
        two
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchResults layout Snapshot on smartphone screens are matched 1`] = `
<div
  className="container-0-2-2"
>
  <div
    className="sidebar-0-2-7"
  >
    <span>
      sidebar here
    </span>
  </div>
  <div
    className="mainContent-0-2-4"
  >
    <div
      className="searchBar-0-2-6"
    >
      <span>
        searchbar here
      </span>
    </div>
    <div
      className="resultsMessage-0-2-5"
    >
      <span>
        Result message here
      </span>
    </div>
    <div
      className="grid-0-2-3"
    >
      <div
        className="cardContainer-0-2-1"
      >
        one
      </div>
      <div
        className="cardContainer-0-2-1"
      >
        two
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchResults layout Snapshot on tablet screens are matched 1`] = `
<div
  className="container-0-2-2"
>
  <div
    className="sidebar-0-2-7"
  >
    <span>
      sidebar here
    </span>
  </div>
  <div
    className="mainContent-0-2-4"
  >
    <div
      className="searchBar-0-2-6"
    >
      <span>
        searchbar here
      </span>
    </div>
    <div
      className="resultsMessage-0-2-5"
    >
      <span>
        Result message here
      </span>
    </div>
    <div
      className="grid-0-2-3"
    >
      <div
        className="cardContainer-0-2-1"
      >
        one
      </div>
      <div
        className="cardContainer-0-2-1"
      >
        two
      </div>
    </div>
  </div>
</div>
`;
